keps/sig-storage/20200127-recover-from-resize.md

```diff
@@ -0,0 +1,175 @@
+
+In practice above limitation should be fine. We will add events to PV and PVC when we detect that user has reduced requested size. Admin can recover from this state either by manually increasing quota or by manually rebuilding the PVC object.
+
+#### User flow stories
+
+###### Case 1 (controller+node expandable):
+
+- User increases 10Gi PVC to 100Gi by changing - `pvc.spec.resources.requests["storage"] = "100Gi"`.
+- Api Server sets `pvc.spec.allocatedResources` to 100Gi via pvc storage strategy.
+- Quota controller uses `max(pvc.Spec.AllocatedResources, pvc.Spec.Resources)` and adds `90Gi` to used quota.
+- Expansion to 100Gi fails and hence `pv.Spec.Capacity` and `pvc.Status.Capacity `stays at 10Gi.
+- User requests size to 20Gi.
+- Quota controller sees no change in storage usage by the PVC because `pvc.Spec.AllocatedResources` is `100Gi`.
+- Expansion succeeds and `pvc.Status.Capacity` and `pv.Spec.Capacity` report new size as `20Gi`.
+- `pvc.Spec.AllocatedResources` however keeps reporting `100Gi`.
+
+
+###### Case 2 (controller+node expandable):
+
+- User increases 10Gi PVC to 100Gi by changing - `pvc.spec.resources.requests["storage"] = "100Gi"`
+- Api Server sets `pvc.spec.allocatedResources` to 100Gi via pvc storage strategy.
+- Quota controller uses `max(pvc.Spec.AllocatedResources, pvc.Spec.Resources)` and adds `90Gi` to used quota.
+- Expansion to 100Gi fails and hence `pv.Spec.Capacity` and `pvc.Status.Capacity `stays at 10Gi.
+- User requests size to 20Gi.
+- Quota controller sees no change in storage usage by the PVC because `pvc.Spec.AllocatedResources` is `100Gi`.
+- Expansion succeeds and `pvc.Status.Capacity` and `pv.Spec.Capacity` report new size as `20Gi`.
+- `pvc.Spec.AllocatedResources` however keeps reporting `100Gi`.
+- User expands PVC to 120Gi
+- Api Server sets `pvc.spec.allocatedResources` to 120Gi.
+- Quota controller uses `max(pvc.Spec.AllocatedResources, pvc.Spec.Resources)` and adds `20Gi` to used quota.
+- Expansion to 120Gi fails and hence `pv.Spec.Capacity` and `pvc.Status.Capacity `stays at 20Gi.
+
+###### Case 3
+
+- User increases 10Gi PVC to 100Gi by changing `pvc.spec.resources.requests["storage"] = "100Gi"`
+- API server sets pvc.spec.allocatedResources to 100Gi via PVC storage strategy.
+- Quota controller uses max(pvc.Spec.AllocatedResources, pvc.Spec.Resources) and adds 90Gi to used quota. (100Gi is the total space taken by the PVC)
+- Volume plugin starts slowly expanding to 100Gi. pv.Spec.Capacity and pvc.Status.Capacity stays at 10Gi until the resize is finished.
+- While the storage backend is resizing the volume, user requests size 20Gi by changing pvc.spec.resources.requests["storage"] = "20Gi"
+- Quota controller sees no change in storage usage by the PVC because pvc.Spec.AllocatedResources is 100Gi.
+- Expansion succeeds and pvc.Status.Capacity and pv.Spec.Capacity report new size as 100Gi, as that's what the volume plugin did.
+
```

Comment:
Isnt there a fourth case where the expansion happend in the storage backend, but fails on second step on node which is volume fs expansion fails on pod spawn meanwhile user changed/shrink the size to less capacity?  or did I miss anything ?

Reply:
I can add that as a fourth case but it does not affect the flow I was trying to illustrate, so I skipped.